### PR TITLE
Add CassandraOperation::queries method to abstract over QUERY and BATCH

### PR DESCRIPTION
**Pull request by rukai**
Progress towards https://github.com/shotover/shotover-proxy/issues/563
I ended up going for a different design then what I described in the issue.
QUERY and BATCH diverge too much to take that route.
So instead I propose we just add a common getter method so we can pretend batches and queries are the same thing in most cases but then fall back to directly accessing query/batch specific fields if we ever need it.

TODO in follow up PRs:
* CassandraOperation::Query should store a single Statement instead of a `Vec<Statement>` (we can use https://doc.rust-lang.org/std/slice/fn.from_mut.html to turn it into a `&mut [T]`)
* Return CassandraOperation::Batch queries in CassandraOperation::queries once we make use of https://docs.rs/cassandra-protocol/latest/cassandra_protocol/frame/frame_batch/struct.BodyReqBatch.html

We can land this PR as is if we are ok with the CassandraOperation::queries API introduced by the PR and can confirm that its use in transforms is sound.

When thinking about maintaining transform invariants when supporting batch requests remember that batch requests only support INSERT/UPDATE/DELETE queries so we dont have to worry about them returning values. (But they could still return an error)

### CODE REVIEW

**Comment by conorbros** [comment]
Is BATCH parsing not already in cassandra-protocol?

**Comment by rukai** [comment]
Oh, turns out it is already implemented: https://docs.rs/cassandra-protocol/latest/cassandra_protocol/frame/frame_batch/struct.BodyReqBatch.html

I still want to make use of it in a follow up PR though.

**Comment by rukai** [comment]
Hmmm I might have to think about this more, the way the batch protocol mixes prepared and regular queries makes it impossible to expose `Statement`s as a slice without butchering the way we store batch queries.

**Comment by rukai** [comment]
All good, I changed the slice to an iterator and we should be able to expose batch queries just fine with that.

File: shotover-proxy/src/frame/cassandra.rs [hunk 1]
```diff
@@ -286,6 +286,18 @@ pub enum CassandraOperation {
 }
 
 impl CassandraOperation {
+    /// Return all queries contained within CassandaOperation::Query and CassandraOperation::Batch
+    pub fn queries(&mut self) -> std::slice::IterMut<Statement> {
```
Comment (conorbros) [inline shotover-proxy/src/frame/cassandra.rs:290]: Add an import

File: shotover-proxy/src/transforms/redis/cache.rs [hunk 2]
```diff
@@ -72,38 +69,48 @@ impl SimpleRedisCache {
...
-        for message in &mut messages {
-            info!("Received reply from redis cache {:?}", message);
-            // TODO: Translate the redis reply into cassandra
-            *message = Message::from_frame(Frame::Cassandra(CassandraFrame {
+
+        // Replace cass_request messages with cassandra responses in place.
+        // We reuse the vec like this to save allocations.
+        let mut messages_redis_response_iter = messages_redis_response.iter_mut();
+        for cass_request in &mut messages_cass_request {
+            let _redis_response = messages_redis_response_iter.next();
```
Comment (conorbros) [inline shotover-proxy/src/transforms/redis/cache.rs:95]: What's this doing?

File: shotover-proxy/src/frame/cassandra.rs [hunk 3]
```diff
@@ -286,6 +286,18 @@ pub enum CassandraOperation {
...
+        match self {
+            CassandraOperation::Query {
+                query: CQL::Parsed(query),
+                ..
+            } => query.iter_mut(),
+            // TODO: Return CassandraOperation::Batch queries once we add BATCH parsing to cassandra-protocol
```
Comment (conorbros) [inline shotover-proxy/src/frame/cassandra.rs:296]: Are all the TODOs intended to be done in the follow up PRs you mention?

File: shotover-proxy/src/transforms/redis/cache.rs (hunk 2 above)
Comment (rukai) [reply shotover-proxy/src/transforms/redis/cache.rs:95]: Its demonstrating how to get the redis responses that correspond to this cassandra request.
Each cassandra request can have multiple redis responses due to batches getting split up into multiple redis messages.

There is nothing to actually use the redis responses as even before this PR this was left as a TODO: `// TODO: Translate the redis reply into cassandra`

File: shotover-proxy/src/frame/cassandra.rs (hunk 3 above)
Comment (rukai) [reply shotover-proxy/src/frame/cassandra.rs:296]: This is the only new TODO and yes I intend to do it in a follow up PR

Comment (conorbros) [reply shotover-proxy/src/frame/cassandra.rs:297]: Oh true, the other one is just whitespace.

File: shotover-proxy/src/transforms/redis/cache.rs (hunk 2 above)
Comment (conorbros) [reply shotover-proxy/src/transforms/redis/cache.rs:95]: Could you add a comment explaining this, all the iterators are a bit hard to follow.

Comment (rukai) [reply shotover-proxy/src/transforms/redis/cache.rs:113]: In order to understand what the iterators are doing you need to understand how the earlier vecs are constructed.
So I wrote some comments explaining my mental model of how the whole function works.
Does this help?

Also, thanks to writing out these comments I noticed and fixed a bug earlier in this function.

File: shotover-proxy/tests/cassandra_int_tests/basic_driver_tests.rs [hunk 4]
```diff
@@ -964,6 +964,7 @@ mod functions {
             "CREATE TABLE test_function_keyspace.test_function_table (id int PRIMARY KEY, x int, y int);",
         );
         run_query(
+            // I think these are sending as query messages lolololol
```
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/basic_driver_tests.rs:967]: oh right this.
I'll investigate.

File: shotover-proxy/src/transforms/redis/cache.rs (hunk 2 above)
Comment (conorbros) [reply shotover-proxy/src/transforms/redis/cache.rs]: Yep that clears it up, thanks!

**Review by conorbros — APPROVED** [review]

**Review by benbromhead — APPROVED** [review]
lgtm
